docs/USAGE-GUIDE.md

```diff
@@ -1388,6 +1388,30 @@ Return the certificate that was used during the TLS handshake.
 This function returns NULL if the certificate selection phase of the handshake has not completed
  or if a certificate was not requested by the peer.
 
+### s2n\_get\_cert\_chain\_length
+
+```c
+int s2n_get_cert_chain_length(struct s2n_cert_chain_and_key *chain_and_key, uint32_t *cert_length);
+```
+
+**s2n_get_cert_chain_length** gets the length of the certificate chain `chain_and_key`. If the certificate chain `chain_and_key` is NULL an error is thrown. If the certificate chain contains only one ceritificate the length returned is equal to one. 
+
+### s2n\_get\_cert\_from\_cert\_chain
+
+```c
+int s2n_get_cert_from_cert_chain(struct s2n_cert_chain_and_key *chain_and_key, struct s2n_cert **out_cert, uint32_t cert_idx);
+```
+
+**s2n_get_cert_from_cert_chain** gets the certificate `out_cert` present at the index `cert_idx` of the certificate chain `chain_and_key`.  If the certificate chain `chain_and_key` is NULL or the certificate index value is not in the acceptable range for the input certificate chain, an error is thrown. Note that the index of the head_cert is zero.
+
+### s2n\_get\_cert\_der
+
+```c
+int s2n_get_cert_der(struct s2n_cert *cert, uint8_t **out_cert_der, uint32_t *cert_length);
+```
+
+**s2n_get_cert_der** gets the certificate `cert` in .der format which is returned in the buffer `out_cert_der`, `cert_len` represents the length of the certificate. 
```

Comment:
If we want to return bytes allocated by S2N, we need to be very clear about how they can be used safely. I assume that they will only be valid as long as the connection lives?

If you don't want to worry about the lifetime of the memory and want to remove that particular foot gun, instead accept memory from the customer and write the certificate to it.

Reply:
Yes, the memory would valid only as long as the connection lives here.

Reply:
Yes, the memory would be valid only as long as the connection lives here.